packages/flutter/lib/src/rendering/binding.dart

```diff
@@ -335,19 +447,7 @@ mixin RendererBinding on BindingBase, ServicesBinding, SchedulerBinding, Gesture
 
   @override
   void performSemanticsAction(SemanticsActionEvent action) {
-    _pipelineOwner.semanticsOwner?.performAction(action.nodeId, action.type, action.arguments);
-  }
-
-  void _handleSemanticsOwnerCreated() {
-    renderView.scheduleInitialSemantics();
-  }
-
-  void _handleSemanticsUpdate(ui.SemanticsUpdate update) {
-    renderView.updateSemantics(update);
-  }
-
-  void _handleSemanticsOwnerDisposed() {
-    renderView.clearSemantics();
+    _viewIdToRenderView[action.viewId]?.owner?.semanticsOwner?.performAction(action.nodeId, action.type, action.arguments);
```

Comment:
Do you want to assert that the `viewId` exists in the `_viewIdToRenderView` map? I mean, it's probably OK to just ignore the action if the view isn't in the map, but it also seems like a programming error if the view isn't in the map.

Reply:
If you decide to update this, don't forget to update hit testing below

Reply:
This is purposely a little forgiving. There is a lot of asynchronicity in screen readers when they process Flutter's semantics tree and send events back to us. Sometimes, when screen readers haven't processed the latest tree update yet, they will send us events for nodes that don't exist anymore (real life issue that caused crashes because of this: https://github.com/flutter/flutter/issues/100358). That's why the action handling code has always been a little more forgiving, always assuming that the screen reader may operate on outdated data, see also our existing implementation for semantics action handling:

https://github.com/flutter/flutter/blob/7e2f1f5bf6e19274e87d7184b6f7249d629d1b18/packages/flutter/lib/src/rendering/binding.dart#L337-L339

https://github.com/flutter/flutter/blob/eebb1d6ddf075b5fee8a93fc00ce3e0cebe8f12c/packages/flutter/lib/src/semantics/semantics.dart#L3341-L3352

Not asserting here and gracefully ignoring an action if the associated view doesn't exist in Flutter anymore is continuing this tradition.

Reply:
Thanks for the explanation! This might be worth a quick comment if this isn't is a well known pattern

Reply:
Does Dart allow simply `_viewIdToRenderView[action.viewId]?.owner.semanticsOwner.performAction` instead? I vaguely remember Dart automatically expands the first `?` to the rest of the chain.